broadcaster/backlog/backlog.go

```diff
@@ -279,48 +265,72 @@ func newBacklogSegment() *backlogSegment {
 // variable of type BacklogSegment is nil or not. Comparing whether an
 // interface is nil directly will not work.
 func IsBacklogSegmentNil(segment BacklogSegment) bool {
-	return segment.(*backlogSegment) == nil
+	if segment == nil {
+		return true
+	} else if segment.(*backlogSegment) == nil {
+		return true
+	}
+	return false
 }
 
 // Start returns the first message index within the backlogSegment.
 func (s *backlogSegment) Start() uint64 {
-	return uint64(s.start.Load())
+	s.messagesLock.RLock()
+	defer s.messagesLock.RUnlock()
+	if len(s.messages) > 0 {
+		return uint64(s.messages[0].SequenceNumber)
+	}
+	return uint64(0)
 }
 
 // End returns the last message index within the backlogSegment.
 func (s *backlogSegment) End() uint64 {
-	return uint64(s.end.Load())
+	s.messagesLock.RLock()
+	defer s.messagesLock.RUnlock()
+	c := len(s.messages)
+	if c == 0 {
+		return uint64(0)
+	}
+	return uint64(s.messages[c-1].SequenceNumber)
 }
 
 // Next returns the next backlogSegment.
 func (s *backlogSegment) Next() BacklogSegment {
-	return s.nextSegment.Load()
+	next := s.nextSegment.Load()
+	if next == nil {
+		return nil // return a nil interface instead of a nil *backlogSegment
+	}
+	return next
 }
 
 // Messages returns all of the messages stored in the backlogSegment.
 func (s *backlogSegment) Messages() []*m.BroadcastFeedMessage {
 	s.messagesLock.RLock()
 	defer s.messagesLock.RUnlock()
-	return s.messages
+	tmp := make([]*m.BroadcastFeedMessage, len(s.messages))
+	copy(tmp, s.messages)
+	return tmp
 }
 
-// get reads messages from the given start to end message index.
-func (s *backlogSegment) get(start, end uint64) ([]*m.BroadcastFeedMessage, error) {
+// Get reads messages from the given start to end message index.
+func (s *backlogSegment) Get(start, end uint64) ([]*m.BroadcastFeedMessage, error) {
+	s.messagesLock.RLock()
+	defer s.messagesLock.RUnlock()
 	noMsgs := []*m.BroadcastFeedMessage{}
-	if start < s.start.Load() {
+	if start < s.Start() {
 		return noMsgs, errOutOfBounds
 	}
 
-	if end > s.end.Load() {
+	if end > s.End() {
 		return noMsgs, errOutOfBounds
 	}
 
-	startIndex := start - s.start.Load()
-	endIndex := end - s.start.Load() + 1
+	startIndex := start - s.Start()
+	endIndex := end - s.Start() + 1
 
-	s.messagesLock.RLock()
-	defer s.messagesLock.RUnlock()
-	return s.messages[startIndex:endIndex], nil
+	tmp := make([]*m.BroadcastFeedMessage, len(s.messages))
+	copy(tmp, s.messages)
+	return tmp[startIndex:endIndex], nil
```

Comment:
I doubt this would matter too much but ideally we'd create tmp to only be the size of endIndex - startIndex and copy that section

Reply:
I actually tried to do that with:
```
tmp := make([]*m.BroadcastFeedMessage, endIndex-startIndex)
copy(tmp, s.messages[startIndex:endIndex])
return tmp, nil
```
but for some reason it seemed to copy messages from the `s.messages` slice outside of the `startIndex` to `endIndex` range that I had specified. I'm not sure why and maybe I did something wrong. For now I will leave this as is as it fixed the errors I saw in the test.

Reply:
actually I must have tried something else when copying just that section of the slice as I added it and it is passing the tests. I think earlier I might have used a different length than `endIndex-startIndex` when making the tmp slice